src/main/java/com/example/test/MainActivity.java

```diff
@@ -1,38 +1,124 @@
+        Intent intent = new Intent(this, MyService.class);
+        bindService(intent, mConnection, Context.BIND_AUTO_CREATE);
+
+        //RecyclerView creation
+        mNumberList = findViewById(R.id.main_RecyclerView);
+        mNumberList.setHasFixedSize(true);
+        LinearLayoutManager NumberListManager = new LinearLayoutManager(this);
+        mNumberList.setLayoutManager(NumberListManager);
 
-        start = (Button) findViewById(R.id.buttonStart);
 
-        stop = (Button) findViewById(R.id.buttonStop);
+        //Buttons creation
+        mGenerateButton = (Button) findViewById(R.id.main_GenerateButton);
+        mSortButton = (Button) findViewById(R.id.main_SortButton);
 
-        start.setOnClickListener(this);
+        mGenerateButton.setOnClickListener(this);
+        mSortButton.setOnClickListener(this);
+
+        //Spinner creation
+        Spinner mSpinner = findViewById(R.id.main_Spinner);
+
+        List<String> captionList = new ArrayList<>();
+        for (Sorter.SortingMethod sort : Sorter.SortingMethod.values()) {
+            captionList.add(getString(sort.getCaptionStringId()));
+        }
 
-        stop.setOnClickListener(this);
+        ArrayAdapter<String> spinnerAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, captionList);
+        spinnerAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
+        mSpinner.setAdapter(spinnerAdapter);
+        mSpinner.setOnItemSelectedListener(this);
     }
 
+    private ServiceConnection mConnection = new ServiceConnection() {
+        @Override
+        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
+            MyService.MyBinder binder = (MyService.MyBinder) iBinder;
+            mBinder = binder;
+            mIsBounded = true;
+        }
+
+        @Override
+        public void onServiceDisconnected(ComponentName componentName) {
+            mIsBounded = false;
+        }
+    };
 
     @Override
     public void onClick(View view) {
-        if (view == start) {
-            startService(new Intent(this, MyService.class));
-        } else if (view == stop) {
-            stopService(new Intent(this, MyService.class));
+        if (view == mGenerateButton) {
+            // 1 . get array from the service () using binder
+            // 2. set it for the adapter
+            mNumbersAdapter = new NumbersAdapter(SIZE_OF_RECYCLER_ELEMENTS);
+            mNumberList.setAdapter(mNumbersAdapter);
+
+            mNumbersAdapter.setDataSet(mBinder.generate(NUMBER_OF_ELEMENTS));
+
+
+        } else if (view == mSortButton) {
+            // 1. check if array is generated
+            // 2. check if mSortingMethod != null
+            if (mNumbersAdapter.getDataSet().length != 0) {
+                if (mSortingMethod != null) {
+                    mNumbersAdapter.setDataSet(mBinder.sort(mNumbersAdapter.getDataSet(), mSortingMethod));
+                }
+            }
```

Comment:
Consider this:
```suggestion
            if (mNumbersAdapter.getDataSet().length == 0) {
                // show error. (Toast, for example)
                return;
            }
            
            if (mSortingMethod == null) {
                // show error
                return;
            }
            
            mNumbersAdapter.setDataSet(mBinder.sort(mNumbersAdapter.getDataSet(), mSortingMethod));
```

Also, it would be even better to move out all this logic into onSortButtonClicked()

Reply:
mSortingMethod has dafault value and nowhere cannot be null. Should i overflow logic?

Reply:
My change has identical logic as the previous code. There is no "overflow logic" there.